libs/datasets/dataset_utils.py

```diff
@@ -330,48 +425,7 @@ def fill_fields_with_data_source(
 
     Returns: Updated dataframe with requested columns filled from data_source data.
     """
-    new_data = data_source.set_index(index_fields)
-
-    # If no data exists, return all rows from new data with just the requested columns.
-    if not len(existing_df):
-        for column in columns_to_fill:
-            if column not in new_data.columns:
-                new_data[column] = None
-        return new_data[columns_to_fill].reset_index()
-    existing_df = existing_df.set_index(index_fields)
-
-    # Sort indices so that we have chunks of equal length in the
-    # correct order so that we can splice in values.
-    existing_df = existing_df.sort_index()
-    new_data = new_data.sort_index()
-
-    # Build series that point to rows that match in each data frame.
-    existing_df_in_new_data = existing_df.index.isin(new_data.index)
-    new_data_in_existing_df = new_data.index.isin(existing_df.index)
-
-    if not sum(existing_df_in_new_data) == sum(new_data_in_existing_df):
-        print(new_data.loc[new_data_in_existing_df, columns_to_fill])
-        existing_in_new = sum(existing_df_in_new_data)
-        new_in_existing = sum(new_data_in_existing_df)
-        raise ValueError(
-            f"Number of rows should be the for data to replace: {existing_in_new} -> {new_in_existing}: {columns_to_fill}"
-        )
-
-    # If a column doesn't exist in the existing data, add it (throws an error)
-    # otherwise.
-    for column in columns_to_fill:
-        if column not in existing_df.columns:
-            existing_df[column] = None
-
-    # Fill in values for rows that match in both data frames.
-    existing_df.loc[existing_df_in_new_data, columns_to_fill] = new_data.loc[
-        new_data_in_existing_df, columns_to_fill
-    ]
-    # Get rows that do not exist in the existing data frame
-    missing_new_data = new_data[~new_data_in_existing_df]
-
-    data = pd.concat(
-        [existing_df.reset_index(), missing_new_data[columns_to_fill].reset_index(),]
+    assert len(columns_to_fill) == 1
```

Comment:
We have interspersed asserts in various parts of the code, but I'd love to move away from them.  Could you maybe make this check raise a `ValueError` if not true?

Reply:
What do you see as the difference between ValueError and AssertionError? They are siblings in https://docs.python.org/3/library/exceptions.html#exception-hierarchy
I see an assert as fine for checking things that are only expected with a fairly local programming error and in a long-running server it is okay to except+log+continue when one raises. ValueError is more likely to be caught due to bad user input, for a very broad interpretation of "user".

Reply:
Good point re `ValueError`.  I've actually never really used `AssertionError` specifically, and probably should consider it more. 

I think my biggest problem with assert is that it doesn't actually run in non-"debug" [modes](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement).  I don't necessarily think that we'll be running the python code with `-o`, but I think I shy away from asserts because of that assumption/reliance on how the python code is executed.
How do you feel about explicitly raising `AssertionError` ?

Reply:
Oh wow, TIL that about `-O`. I thought https://www.python.org/dev/peps/pep-0488/#completely-dropping-optimization-levels-from-cpython was the status-quo and have even commented them out of code to save CPU :-/
Changed to raising `AssertionError`. I plan to delete this function shortly.